### DISCUSSION TITLE
An example of how timer synchronization works

### DISCUSSION THREAD

**Discussion by sadkotheguest** [hpmicro/hpm_sdk#23] [General]
Hello. 
At the moment, I'm trying to figure out how to synchronize timers in HPM6280.

Could you tell me if it is possible to clock two timers connected to the PWM from the third timer? 
Is it possible to synchronously reset timers by an external event (TRGI)?

**Comment by sadkotheguest** [comment]
According to HPM6200_UM_V1_0.
34.2.1 Timer time reference

> 用户也可以利用 PWM 定时器的同步触发输入（SYNCI）来同步计数器开始计数的时机。用户可以配置GCR[RLDSYNCEN] 和 GCR [XRLDSYNCEN] 位来选择是否打开外部同步。GCR [RLDSYNCEN] 位置 1 时，同步触发输入（SYNCI）同步触发输入可以重置计数器到起始值（CNT = STA），重载标志位RLDF置 1。如果GCR[RLDSYNCEN] 位置 1，计数器的扩展位会充值到起始值的扩展位（XCNT = XSTA），扩展重载标志位XRLD也会置 1。

> Users can also use the synchronous trigger input (SYNCI) of the PWM timer to synchronize the timing when the counter starts counting. The user can configure the GCR [RLDSYNCEN] and GCR [XRLDSYNCEN] bits to choose whether to turn on external synchronization. At GCR [RLDSYNCEN] position 1, the synchronization trigger input (SYNCI) synchronization trigger input can reset the counter to the starting value (CNT = STA), and the overload flag RLDF is set to 1. If GCR [RLDSYNCEN] is set to 1, the extension bit of the counter will be reset to the extension bit of the starting value (XCNT = XSTA), and the extension overload flag XRLD will also be set to 1.

Could you tell - how should I setup the TRGO?

**Comment by sadkotheguest** [comment]
An example of using _trgm_output_update_source_
https://github.com/hpmicro/hpm_sdk/blob/7440098831b16ed0b14e97004441bcabcc3f19b2/samples/drivers/pwm/hrpwm/src/hrpwm.c#L42

```(C)
void reset_pwm_counter(void)
{
    pwm_enable_reload_at_synci(HRPWM);
    trgm_output_update_source(TRGM, TRGM_TRGOCFG_PWM_SYNCI, 1);
    trgm_output_update_source(TRGM, TRGM_TRGOCFG_PWM_SYNCI, 0);
}
```

_TRGM_TRGOCFG_PWM_SYNCI_ = 14 is the TRGMn_OUTPUT MUX SOURCE
Hardcoded '1' and '0' should be the SOURCE of TRGM...

**Comment by sadkotheguest** [comment]
https://github.com/hpmicro/hpm_sdk/blob/7440098831b16ed0b14e97004441bcabcc3f19b2/samples/drivers/sdm/src/sdm.c#L229

`trgm_output_update_source(TEST_SDM_TRGM, TEST_SDM_TRGM_OUTPUT_DST, TEST_SDM_TRGM_INPUT_SRC);`

```
#define BOARD_SDM_TRGM                     HPM_TRGM3

// TRGM_OUTPUT
[TRGM3_OUTPUT MUX]
#define BOARD_SDM_TRGM_OUTPUT_DST          HPM_TRGM3_OUTPUT_SRC_SDFM_TRG15
#define HPM_TRGM3_OUTPUT_SRC_SDFM_TRG15                    (0x43UL)
// in decimal 0x43UL = 67 -> SDFM_TRG15 (Table 176: TRGM3 output MUXEL)

// TRGM_INPUT
[TRGM3_INPUT MUX]
#define BOARD_SDM_TRGM_INPUT_SRC           HPM_TRGM3_INPUT_SRC_GPTMR3_OUT2
#define HPM_TRGM3_INPUT_SRC_GPTMR3_OUT2                    (0x26UL)
// in decimal 0x26UL = 38 -> GPTMR3_OUT2 (Table 172: TRGM3 input MUXEL)

```

**Comment by willChuai** [comment]
Yes, it is possible to synchronise other PWMs with the sync timer. The schematic code is as follows, synchronising the four PWM0-PWM3 RLDs. 


```

#define SYNC_TIME_0 (800U)
#define SYNC_TIME_1 (1600U)
#define SYNC_TIME_2 (2400U)
#define SYNC_TIME_3 (3200U)


void init_pwm_sync(SYNT_Type *ptr)
{
    synt_enable_counter(ptr, false);
    synt_reset_counter(ptr);
    synt_set_reload(ptr, 0xffffffff);
    synt_set_comparator(ptr, SYNT_CMP_0, SYNC_TIME_0);
    synt_set_comparator(ptr, SYNT_CMP_1, SYNC_TIME_1);
    synt_set_comparator(ptr, SYNT_CMP_2, SYNC_TIME_2);
    synt_set_comparator(ptr, SYNT_CMP_3, SYNC_TIME_3);
    synt_enable_counter(ptr, true);
}


void pwm_sync_init(void)
{
    trgm_output_t config;

    config.invert = false;
    config.input = HPM_TRGM0_INPUT_SRC_SYNT0_CH0;
    config.type = trgm_output_same_as_input;
    trgm_output_config(HPM_TRGM0, TRGM_TRGOCFG_PWM_SYNCI, &config);

    config.invert = false;
    config.input = HPM_TRGM1_INPUT_SRC_SYNT_CH1;
    config.type = trgm_output_same_as_input;
    trgm_output_config(HPM_TRGM1, TRGM_TRGOCFG_PWM_SYNCI, &config);

    config.invert = false;
    config.input = HPM_TRGM2_INPUT_SRC_SYNT_CH2;
    config.type = trgm_output_same_as_input;
    trgm_output_config(HPM_TRGM2, TRGM_TRGOCFG_PWM_SYNCI, &config);

    config.invert = false;
    config.input = HPM_TRGM3_INPUT_SRC_SYNT_CH3;
    config.type = trgm_output_same_as_input;
    trgm_output_config(HPM_TRGM3, TRGM_TRGOCFG_PWM_SYNCI, &config);

    init_pwm_sync(HPM_SYNT);
}

void reset_pwm_counter(PWM_Type *ptr)
{
    pwm_enable_reload_at_synci(ptr);
}
void pwm_sync(void)
{
    reset_pwm_counter(HPM_PWM0);
    reset_pwm_counter(HPM_PWM1);
    reset_pwm_counter(HPM_PWM2);
    reset_pwm_counter(HPM_PWM3);
    pwm_sync_init();
}


```

**Reply by sadkotheguest** [reply]
Thanks for your help.
